File: mobile/packages/mail/android/src/main/kotlin/io/ente/mail/MailPlugin.kt
package io.ente.mail

import android.app.Activity
import io.ente.mail.core.MailAttachment
import io.ente.mail.core.MailComposer
import io.ente.mail.core.MailRequest
import io.ente.mail.core.MailResult
import io.ente.mail.core.MailUnavailableReason
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class MailPlugin : FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var composer: MailComposer
    private var activity: Activity? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        composer = MailComposer(binding.applicationContext)
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method != "compose") {
            result.notImplemented()
            return
        }
        val request = try {
            call.toRequest()
        } catch (error: IllegalArgumentException) {
            result.error("invalidDraft", error.message, null)
            return
        }
        composer.compose(request, { activity }) { outcome ->
            result.success(outcome.channelValue)
        }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        activity = null
        channel.setMethodCallHandler(null)
        composer.close()
    }

    private companion object {
        const val CHANNEL = "io.ente.mail/composer"
    }
}

private val MailResult.channelValue: Map<String, String>
    get() = when (this) {
        MailResult.Launched -> mapOf("status" to "launched")
        is MailResult.Unavailable -> mapOf(
            "status" to "unavailable",
            "reason" to reason.channelValue
        )
    }

private val MailUnavailableReason.channelValue: String
    get() = when (this) {
        MailUnavailableReason.NO_MAIL_CLIENT -> "noMailClient"
        MailUnavailableReason.ATTACHMENT_COMPOSER_UNAVAILABLE -> {
            "attachmentComposerUnavailable"
        }
        MailUnavailableReason.ATTACHMENT_MISSING -> "attachmentMissing"
        MailUnavailableReason.ATTACHMENT_UNREADABLE -> "attachmentUnreadable"
        MailUnavailableReason.ATTACHMENT_TOO_LARGE -> "attachmentTooLarge"
        MailUnavailableReason.COMPOSER_BUSY -> "composerBusy"
        MailUnavailableReason.PRESENTATION_FAILED -> "presentationFailed"
    }

private fun MethodCall.toRequest(): MailRequest {
    val arguments = arguments as? Map<*, *>
        ?: throw IllegalArgumentException("Mail draft is missing")
    fun string(name: String): String = arguments[name] as? String
        ?: throw IllegalArgumentException("Mail draft has no $name")

    val attachment = arguments["attachment"]?.let { value ->
        val map = value as? Map<*, *>
            ?: throw IllegalArgumentException("Attachment is invalid")
        fun attachmentString(name: String): String = map[name] as? String
            ?: throw IllegalArgumentException("Attachment has no $name")

        MailAttachment(
            path = attachmentString("path"),
            mimeType = attachmentString("mimeType")
        )
    }
    return MailRequest(
        recipient = string("recipient"),
        subject = string("subject"),
        body = string("body"),
        attachment = attachment
    )
}

File: mobile/packages/mail/ios/Classes/MailPlugin.swift
@preconcurrency import Flutter
import Foundation
import UIKit

@MainActor
public final class MailPlugin: NSObject, @preconcurrency FlutterPlugin {
    private let composer = MailComposer()
    private var isDetached = false

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "io.ente.mail/composer",
            binaryMessenger: registrar.messenger()
        )
        let instance = MailPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.publish(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard call.method == "compose" else {
            result(FlutterMethodNotImplemented)
            return
        }
        let request: MailRequest
        do {
            request = try call.mailRequest()
        } catch {
            result(
                FlutterError(
                    code: "invalidDraft",
                    message: error.localizedDescription,
                    details: nil
                )
            )
            return
        }
        guard let presenter = UIApplication.shared.activePresenter else {
            result(MailResult.unavailable(.presentationFailed).channelValue)
            return
        }

        Task { [weak self] in
            guard let self else { return }
            let value = await composer.compose(request, presenter: presenter).channelValue
            guard !isDetached else { return }
            result(value)
        }
    }

    public func detachFromEngine(for _: FlutterPluginRegistrar) {
        isDetached = true
        composer.close()
    }
}

private extension UIApplication {
    var activePresenter: UIViewController? {
        connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .filter { $0.activationState == .foregroundActive }
            .flatMap(\.windows)
            .first(where: \.isKeyWindow)?
            .rootViewController
    }
}

private extension FlutterMethodCall {
    func mailRequest() throws -> MailRequest {
        guard let arguments = arguments as? [String: Any] else {
            throw MailValidationError("Mail draft is missing")
        }
        func string(_ name: String) throws -> String {
            guard let value = arguments[name] as? String else {
                throw MailValidationError("Mail draft has no \(name)")
            }
            return value
        }

        let attachment: MailAttachment?
        if let value = arguments["attachment"] {
            guard let map = value as? [String: Any],
                let path = map["path"] as? String,
                let mimeType = map["mimeType"] as? String
            else {
                throw MailValidationError("Attachment is invalid")
            }
            attachment = try MailAttachment(
                path: path,
                mimeType: mimeType
            )
        } else {
            attachment = nil
        }
        return try MailRequest(
            recipient: string("recipient"),
            subject: string("subject"),
            body: string("body"),
            attachment: attachment
        )
    }
}

private extension MailResult {
    var channelValue: [String: String] {
        switch self {
        case .launched:
            return ["status": "launched"]
        case let .unavailable(reason):
            return ["status": "unavailable", "reason": reason.rawValue]
        }
    }
}
